src/ray/core_worker/lessor_picker.h

```diff
@@ -0,0 +1,73 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "absl/base/thread_annotations.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "ray/gcs/redis_gcs_client.h"
+
+namespace ray {
+
+struct LocalityData {
+  uint64_t object_size;
+  absl::flat_hash_set<NodeID> nodes_containing_object;
+};
+
+/// Interface for mocking the lessor picker.
+class LessorPickerInterface {
+ public:
+  /// Get the address of the best lessor node for the provided task.
+  virtual absl::optional<rpc::Address> GetBestNodeAddrForTask(
+      const TaskSpecification &spec) = 0;
+
+  virtual ~LessorPickerInterface() {}
+};
+
+/// Interface for providers of locality data to the lessor picker.
+class LocalityDataProviderInterface {
+ public:
+  virtual absl::optional<LocalityData> GetLocalityData(const ObjectID &object_id) = 0;
+
+  virtual ~LocalityDataProviderInterface() {}
+};
+
+/// Class used by the core worker to pick the best lessor to request a lease from.
+class LessorPicker : public LessorPickerInterface {
```

Comment:
I'm still not totally set on the name `LessorPicker` for this abstraction. If y'all think of something better, let me know!

Reply:
I think maybe LeasePolicy or something similar with the word "policy", to indicate that this is used for scheduling decisions.